src/history/HistoryArchive.cpp

```diff
@@ -288,22 +288,93 @@ HistoryArchiveState::containsValidBuckets(Application& app) const
     // Level 0 future buckets are always clear
     assert(currentBuckets[0].next.isClear());
 
-    for (uint32_t i = 1; i < BucketList::kNumLevels; i++)
+    uint32_t minBucketVersion = 0;
+    uint32_t oldestBucketVersion = 0;
+    bool nonEmptySeen = false;
+    Hash const emptyHash;
+
+    auto validateBucketVersion = [&](uint32_t bucketVersion) {
+        if (bucketVersion < minBucketVersion)
+        {
+            CLOG(ERROR, "History")
+                << fmt::format("Incompatible bucket versions: expected version "
+                               "{} or higher, got {}",
+                               minBucketVersion, bucketVersion);
+            return false;
+        }
+        return true;
+    };
+
+    // Process bucket, return version
+    auto getVersion = [&](std::string const& bucketHash) {
+        auto bucket =
+            app.getBucketManager().getBucketByHash(hexToBin256(bucketHash));
+        assert(bucket);
+        int32_t version = 0;
+        if (bucket->getHash() != emptyHash)
+        {
+            version = Bucket::getBucketVersion(bucket);
+            if (!nonEmptySeen)
+            {
+                nonEmptySeen = true;
+                oldestBucketVersion = version;
+            }
+        }
+        return version;
+    };
+
+    // Iterate bottom-up, from oldest to newest buckets
+    for (uint32_t i = BucketList::kNumLevels - 1; i > 0; i--)
```

Comment:
I think this loop can be changed to something easier to understand:
you already have a function to process a bucket `getVersion` (that should really be called `processBucket` as it has side effects), that also does `validateBucketVersion`, updates `minBucketVersion` and perform the check against `Bucket::FIRST_PROTOCOL_SHADOWS_REMOVED + 1`.

Then you just end up with a loop:
```
for (uint32_t i = BucketList::kNumLevels - 1; i > 0; i--)
{
   auto const& level = currentBuckets[i];
   if (!processBucket(level.snap) || !processBucket(level.cur))
   {
        return false;
   }
}
```
Note that you could also use `std::accumulate` here and pass around version information but I am not sure it's that much easier to read.

Reply:
Updated the first half of the function to use the loop like you suggested. Note that the check against `Bucket::FIRST_PROTOCOL_SHADOWS_REMOVED + 1` is done in the second half during `next` validation, so I kept it there.

Reply:
A few questions.

Why is this true?
 >     // By protocol "Bucket::FIRST_PROTOCOL_SHADOWS_REMOVED + 1", all
 >     // buckets must be at least of version "Bucket::FIRST_PROTOCOL_SHADOWS_REMOVED"

Is it that it just happens to be true in pubnet? Wondering about testnet resets...

If we can keep that check, should it just be done after the loop as `minBucketVersion` covers everything?

```
        // By protocol "Bucket::FIRST_PROTOCOL_SHADOWS_REMOVED + 1", all
        // buckets must be at least of version "Bucket::FIRST_PROTOCOL_SHADOWS_REMOVED"
        if (minBucketVersion >= Bucket::FIRST_PROTOCOL_SHADOWS_REMOVED + 1 &&        
                oldestBucketVersion < Bucket::FIRST_PROTOCOL_SHADOWS_REMOVED)
        {
            CLOG_ERROR(History,
                        "Invalid HAS: bucket version {} is too old",
                        oldestBucketVersion);
            return false;
        }
```

The previous code was skipping checks when `snap` was the empty bucket, the new code computes `prevSnapVersion` in that case that returns `0`.
When this happens, are we sure that `level.next.hasOutputHash()` is true? Is it possible that we have instead `level.next.isClear()` ?

Reply:
> Is it that it just happens to be true in pubnet? Wondering about testnet resets...

This check intended to validate older buckets (specifically, version 12 and 13). During testnet reset, we start generating new buckets with the latest protocol version. For example, right now the bucketlist at ledger 1000 on testnet has some version 15 buckets, and the rest are empty buckets (satisfying this check). This code would be invalid for new networks that start at some older protocol (for example, 10), then do multiple protocol upgrades within a short period of time. I assumed that was not a valid scenario. What do you think? 

> If we can keep that check, should it just be done after the loop as minBucketVersion covers everything?

I placed it there to fail fast in case the version is bad, to avoid unnecessary file reads from disk (that's how we get the bucket version).

> The previous code was skipping checks when snap was the empty bucket, the new code computes prevSnapVersion in that case that returns 0.
When this happens, are we sure that level.next.hasOutputHash() is true? Is it possible that we have instead level.next.isClear() ?

I _believe_ so. Prior to protocol 12, we always resolved merges-in-progress. Note that in case both inputs are empty buckets, an empty output bucket is still produced. This can be verified with some really old pubnet ledgers, as there were lots of empty buckets then. We've started actively clearing `next` in protocol 12. The only other scenario where we'd end up with pre-protocol-12 bucket levels that have `next` cleared is when the network just started closing ledgers, and some levels are brand new and haven't been filled yet. This is guarded by the `nonEmptySeen` check.

But there can never be too much of bucketlist paranoia! :) So maybe @graydon can take a look at my reasoning above as well?  ^^

Reply:
> This check intended to validate older buckets (specifically, version 12 and 13). During testnet reset, we start generating new buckets with the latest protocol version

For this to be true we'd have to guarantee that no buckets are generated at protocol 0, right? I think this should be fine. I am going to double check the script that we use.
This would potentially bust private networks though? I think we should add a note somewhere about this.
Let's add a note at the end of the admin guide: a new section "Considerations when bootstrapping a new network", where we can explain that people need to upgrade right away to the latest version of the protocol before accepting any transactions, and provide the command to run to do it.

> I placed it there to fail fast in case the version is bad, to avoid unnecessary file reads from disk (that's how we get the bucket version).

I don't think fast failing here is that important, we're talking a few ms difference. I think the code is easier to understand when broken up: when I moved it in my local copy, I had then more questions on the main validation code that then fits easily on the screen.